### ISSUE TITLE
Review and Recommendations for Exception handling

### ISSUE THREAD

**Issue by vchendrix** [BASIN-3D/basin3d#1]

**Comment by vchendrix** [comment]
The exception handling mechanision centers mostly on functionality for data synthesis calles (MonitoringFeature, MeasurementTimeseriesTVObservation).  When implementing exceptions handling other BASIN-3D functions should be reviewed  for needs there. The goal here is implement exception handling and providing helpful messages to the : 

- Plugin users
- Plugin developers

This does not include logging.  That design will be addressed in #90

The BASIN-3D framework will provide that classes and data structures for the plugin developers to implement in their plugins.  This design will return errors and warning for a synthesis query. 

- **Datasource synthesis error:**  this is an error processing the data. Plugin developers will be responsible for raising these for the most part. All thrown exceptions will fall into this category.
- **Datasource synthesis warning**: Plugin developer is responsible for these. Warnings are passed back at the end of the data synthesis with StopIteration(warnings).  
- **Datasource error:** general errors with the data source. These could be errors such as invalid credentials, datasource endpoints not accessible or something lile that. BASIN-3D framework should be responsible for this, I think.
- **Datasource warning:** this could be a message to the user saying the data source does not implement the requested model or that a feature type is not supported. BASIN-3D framework should be responsible for this. 
- **Datasource Plugin Error:** these could be configuration errors with the plugin in general and the user needs to be pointed to contacting the plugin developers.  This could also be a mechanism for debugging a plugin.   BASIN-3D framework should be responsible for handling this.

Notes:

- Checking for plugin supported features types should be handled outside of the plugins. This would remove the need to log a warning inside the plugins.
- Need to handle duplication of messages.  Be mindful that the same warning message are not duplicated many times. BASIN-3D should limit the number of warning messages allowed. If a plugin developer wants to provide messages for the entiring processing these should be done through logging.

Extend `SynthesisReponse` to have a messages field. Using pydantic error handling as an inspiration (https://pydantic-docs.helpmanual.io/usage/models/#error-handling).

```
SynthesisResponse
   query
   result
   messages: List<Message>]
```

Add a new `SynthesisMessage` data structure

```
SyntheisMessage
   level: WARN, ERROR, CRITICAL
   loc: List[str] - If empty or null, this is a BASIN-3D error, the first item in the list is the datsource id, the second should be the synthesis model. Not sure if there should be subsequent nesting.
   msg
```

**Example JSON output**

```json
{
  “messages”: [
      {“level”: “”, “loc”: [<datasource>, <model>], “msg”: “”},
     ],
  “query”: ...,
  “results”: ...
}
```

**Comment by dschristianson** [comment]
@vchendrix: Looks great. Thank you!
